### PR TITLE
core: fix queued transaction eviction

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR is the successor of #21119 by dropping the `beats` field in txpool.

@holiman @villanuevawill Please take a look if the PR is good to you guys.

**If we want to merge this PR, please don't squash into a single commit, would be nice to keep the commit from vill**

### CODE REVIEW

File: core/tx_pool.go [hunk 1]
```diff
@@ -1358,6 +1371,7 @@ func (pool *TxPool) truncateQueue() {
 			}
 			drop -= size
 			queuedRateLimitMeter.Mark(int64(size))
+			delete(pool.beats, addr.address)
```
Comment (holiman) [inline core/tx_pool.go:1374]: I don't understand this. First of all, won't the `pool.removeTx(tx.Hash(),true)` internally 'fix' the `pool.beats`, if needed? And secondly, isn't it possible that you here will delete the address from beats even though not all transactions were removed?

Comment (rjl493456442) [reply core/tx_pool.go:1374]: Yes, you are right `removeTx` will drop the `pool.beats`. 

> And secondly, isn't it possible that you here will delete the address from beats even though not all transactions were removed?

In this clause, we will drop all transactions, so in theory the queue will be empty.

But anyway i will remove it. Thanks for catching it.

**Comment by villanuevawill** [comment]
@rjl493456442  I'd be curious to understand the progression here. This is mostly identical to the initial solution I included, however, this only updates the beat when the first transaction is queued (on promotion as well). So if an account has no promotions for the lifetime, all new queued transactions will be ejected based on when the first queue item was included. Is this the intended behavior? If we have no recent promotions, this is essentially what will happen now:

1. tx1 added to the queue
2. tx2 added to the queue just before the config.lifetime duration has passed
3. config.lifetime duration passes and both transactions are evicted
4. tx3 added again to the queue right after eviction (stays for the config.lifetime duration)

This **does** fix the initial issue, but doesn't seem to be as consistent in its behavior.
re: conversations here:
https://github.com/ethereum/go-ethereum/pull/21119#issuecomment-638652449
https://github.com/ethereum/go-ethereum/pull/21119#issuecomment-645610771

**Comment by holiman** [comment]
Gary and I discussed it a bit, before this PR was made. So the original PR pivoted on this point

>>    For the latter one, I think the best solution may record the timestamp of each transaction(but tbh not sure whether it's worthwhile to add this complexity).
>Yes, this would be ideal.

And then per-tx timestamps were added. And suddenly there was both per-tx timestamp _and_ per-account timestamp, which seemed kind of odd. And then Gary removed the per-account timestamp, to have only the per-tx ones. 

However, the effect of that was that it added a lot of processing, since the tx handling has to flatten the lists and iterate all txs in the loop, instead of just checking one timestamp. So we agreed that maybe it would be better to focus on the original implementation by Peter, which used only one timestamp per account. 

And yeah, we seem to have circled back to your original PR :) (more or less -- they're not identical, but they seem to be very similar).

**Comment by villanuevawill** [comment]
> However, the effect of that was that it added a lot of processing, since the tx handling has to flatten the lists and iterate all txs in the loop, instead of just checking one timestamp. So we agreed that maybe it would be better to focus on the original implementation by Peter, which used only one timestamp per account.

Makes sense - hopefully, I was helpful here. I'll likely put up some other PRs/bugfixes in the near term (looking at the backlog).

**Comment by rjl493456442** [comment]
@villanuevawill Hi, I would like to apologize first for the first suggestions I made. Recording the timestamp for each transaction is not a good idea. After reworking based on your code, I think I understand the original mechanism better so I choose to come back to the initial version(+ some fixes). 

> Makes sense - hopefully, I was helpful here.
We really appreciate your contribution, thanks for digging out the issue and open the initial fix PR!

File: core/tx_pool.go [hunk 2]
```diff
@@ -362,9 +363,12 @@ func (pool *TxPool) loop() {
 				}
 				// Any non-locals old enough should be removed
 				if time.Since(pool.beats[addr]) > pool.config.Lifetime {
-					for _, tx := range pool.queue[addr].Flatten() {
+					list := pool.queue[addr].Flatten()
+					size := len(list)
+					for _, tx := range list {
 						pool.removeTx(tx.Hash(), true)
 					}
+					queuedEvictionMeter.Mark(int64(size))
```
Comment (karalabe) [inline core/tx_pool.go:371]: Nit, please inline `int64(len(list))`. Seems weird to pre-calculate this field and store it. I was expecting something else to be done with it later.

File: core/tx_pool.go [hunk 3]
```diff
@@ -618,6 +622,9 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (replaced bool, err e
 		pool.journalTx(from, tx)
 		pool.queueTxEvent(tx)
 		log.Trace("Pooled new executable transaction", "hash", hash, "from", from, "to", tx.To())
+
+		// Successfully promotion, bump the heartbeat
```
Comment (karalabe) [inline core/tx_pool.go:626]: "Successfully" -> "Successful"

File: core/tx_pool.go [hunk 4]
```diff
@@ -720,9 +729,10 @@ func (pool *TxPool) promoteTx(addr common.Address, hash common.Hash, tx *types.T
 		pool.priced.Put(tx)
 	}
 	// Set the potentially new pending nonce and notify any subsystems of the new tx
-	pool.beats[addr] = time.Now()
 	pool.pendingNonces.set(addr, tx.Nonce()+1)
 
+	// Successfully promotion, bump the heartbeat
```
Comment (karalabe) [inline core/tx_pool.go:734]: "Successfully" -> "Successful"

**Pull request merged**